gradle-junit-reports/src/main/java/com/palantir/gradle/junit/BuildJunitReportService.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * (c) Copyright 2017 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.gradle.junit;
+
+import com.palantir.gradle.junit.BuildJunitReportService.ReportFile;
+import java.io.File;
+import java.io.Serializable;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+import org.gradle.tooling.events.FinishEvent;
+import org.gradle.tooling.events.OperationCompletionListener;
+import org.gradle.tooling.events.task.TaskFailureResult;
+import org.gradle.tooling.events.task.TaskFinishEvent;
+import org.gradle.tooling.events.task.TaskOperationResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.w3c.dom.Document;
+
+public abstract class BuildJunitReportService
+        implements BuildService<ReportFile>, OperationCompletionListener, AutoCloseable {
+    private static final Logger log = LoggerFactory.getLogger(BuildJunitReportService.class);
+    private final List<Report.TestCase> testCases = new ArrayList<>();
```

Comment:
Think we probably want to ensure thread safe access to this field

```suggestion
    @GuardedBy("this")
    private final List<Report.TestCase> testCases = new ArrayList<>();
```

Reply:
could we use a ConcurrentHashSet and drop the guard and explicit synchronization?